### PR TITLE
CN support reference free

### PR DESCRIPTION
**Pull request by Vela-zz**
1. Add readability to chinese module. @yosukehigashi, take a look at your convenience. 
update with latest main branch.

For futre readability func update， [A Unified Neural Network Model for Readability Assessment
with Feature Projection and Length-Balanced Loss](https://arxiv.org/pdf/2210.10305.pdf), this paper descirbe a linear classification model mix BERT embedding with hand-crafted features. None of the datasets/code are released up to now.

### CODE REVIEW

File: src/langcheck/metrics/zh/reference_free_text_quality.py [hunk 1]
```diff
@@ -217,10 +221,96 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
...
             # yapf: disable
             if label_proba['label'] == _model_id2label[0]:  # type: ignore[reportGeneralTypeIssues]  # NOQA: E501
                 toxicity_scores.append(1 - label_proba['score'])  # type: ignore[reportGeneralTypeIssues]  # NOQA: E501
             # yapf: enable
     return toxicity_scores  # type: ignore[reportGeneralTypeIssues]
+
+
+def xuyaochen_report_readability(
+        generated_outputs: List[str]) -> MetricValue[float]:
+    '''Calculates the readibility scores of generated outputs introduced in
```
Comment (yosukehigashi) [inline src/langcheck/metrics/zh/reference_free_text_quality.py:235]:
```suggestion
    '''Calculates the readability scores of generated outputs introduced in
```

File: src/langcheck/metrics/zh/reference_free_text_quality.py [hunk 2]
```diff
@@ -217,10 +221,96 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
...
+    "中文年报可读性"(Chinese annual report readability). This metrics calcuate
+    average words per sentence as r1, average adverb numbres and connection
```
Comment (yosukehigashi) [inline src/langcheck/metrics/zh/reference_free_text_quality.py:237]: what does "average adverb numbers and connection words" mean?

File: src/langcheck/metrics/zh/reference_free_text_quality.py [hunk 3]
```diff
@@ -217,10 +221,96 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
...
         for label_proba in item_predict_proba:  # type: ignore[reportGeneralTypeIssues]  # NOQA: E501
             # yapf: disable
             if label_proba['label'] == _model_id2label[0]:  # type: ignore[reportGeneralTypeIssues]  # NOQA: E501
                 toxicity_scores.append(1 - label_proba['score'])  # type: ignore[reportGeneralTypeIssues]  # NOQA: E501
             # yapf: enable
     return toxicity_scores  # type: ignore[reportGeneralTypeIssues]
+
+
+def xuyaochen_report_readability(
+        generated_outputs: List[str]) -> MetricValue[float]:
```
Comment (yosukehigashi) [inline src/langcheck/metrics/zh/reference_free_text_quality.py:234]: Let's add `prompts: Optional[List[str] | str] = None` as a parameter here, and then add the prompts to the metric value output. (`flesch_reading_ease` and `tateishi_ono_yamada_reading_ease` have this too)

File: src/langcheck/metrics/zh/reference_free_text_quality.py [hunk 4]
```diff
@@ -217,10 +221,96 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
...
+    '''Calculates the readibility scores of generated outputs introduced in
+    "中文年报可读性"(Chinese annual report readability). This metrics calcuate
+    average words per sentence as r1, average adverb numbres and connection
+    words as r2, then, refer to the Fog Indexto combine the r1 and r2.
+    This function use HanLP Tokenizer and Pos at the same time, POS
+    in CTB style https://hanlp.hankcs.com/docs/annotations/pos/ctb.html。
```
Comment (yosukehigashi) [inline src/langcheck/metrics/zh/reference_free_text_quality.py:240]:
```suggestion
    in CTB style https://hanlp.hankcs.com/docs/annotations/pos/ctb.html.
```

**Review by yosukehigashi — COMMENTED** [review]
Thanks for the change! Mostly LGTM - just a few minor comments

File: src/langcheck/metrics/zh/reference_free_text_quality.py [hunk 5]
```diff
@@ -217,10 +221,99 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
...
+            return 0
+        else:
+            return sum(pos_count_by_sentnece) / len(pos_count_by_sentnece)
+
+    r1 = list(map(calc_r1, output_tokens))   # type: ignore[reportGeneralTypeIssues] # NOQA: E501
+    r2 = list(map(calc_r2, output_pos))   # type: ignore[reportGeneralTypeIssues] # NOQA: E501
+    r3 = [(r1_score + r2_score) * 0.5 for r1_score, r2_score in zip(r1, r2)]
+    # yapf: enable
+    return MetricValue(metric_name='readability',
+                       prompts=None,
```
Comment (yosukehigashi) [inline src/langcheck/metrics/zh/reference_free_text_quality.py:313]:
```suggestion
                       prompts=prompts,
```

File: src/langcheck/metrics/zh/reference_free_text_quality.py (hunk 2 above)
Comment (yosukehigashi) [reply src/langcheck/metrics/zh/reference_free_text_quality.py]: oh interesting, so it's actually just counting the number of adverbs and conjunction words. Interesting that that's how readability is computed!

File: src/langcheck/metrics/zh/reference_free_text_quality.py [hunk 6]
```diff
@@ -217,10 +221,96 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
...
+
+    def calc_r1(content: List[List[str]]) -> float:
+        token_count_by_sentence = list(map(count_tokens, content))
+        if len(token_count_by_sentence) == 0:
+            return 0
+        else:
+            return sum(token_count_by_sentence) / len(token_count_by_sentence)
+
+    def calc_r2(content: List[List[str]]) -> float:
+        pos_count_by_sentnece = list(map(count_postags, content))
```
Comment (yosukehigashi) [inline src/langcheck/metrics/zh/reference_free_text_quality.py:302]:
```suggestion
        pos_count_by_sentence = list(map(count_postags, content))
```
(Missed this before!)

File: src/langcheck/metrics/zh/reference_free_text_quality.py (hunk 2 above)
Comment (Vela-zz) [reply src/langcheck/metrics/zh/reference_free_text_quality.py]: yes, other researcher also propose a method calculates readability based on the character's "画数" (stroke) in a sentence. I think devlope a readability eval algo align with human feeling is difficult and not be fully discussed in research area. In Chinese, the meaning of readability/reading ease seems very similiar with fluency(related with grammar mistakes...).

**Review by yosukehigashi — APPROVED** [review]
I committed two small changes (https://github.com/citadel-ai/langcheck/pull/61/commits/c922daf385a670424df587c6abf1ef641d310583 and https://github.com/citadel-ai/langcheck/pull/61/commits/61c34987a327c37b2c9bdb830e896a7bbb626b9c), so if those look ok to you I think we can go ahead and merge this! Thanks again for the awesome work @Vela-zz!

**Pull request merged**